src/wasmtime/wasmtime.cc

```diff
@@ -0,0 +1,699 @@
+    getModuleFunctionImpl(function_name, f);                                                       \
+  };
+  FOR_ALL_WASM_VM_EXPORTS(_GET_MODULE_FUNCTION)
+#undef _GET_MODULE_FUNCTION
+private:
+  bool getStrippedSource(WasmByteVec *out);
+
+  template <typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                void (*function)(void *, Args...));
+
+  template <typename R, typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                R (*function)(void *, Args...));
+
+  template <typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<void(ContextBase *, Args...)> *function);
+
+  template <typename R, typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<R(ContextBase *, Args...)> *function);
+
+  WasmByteVec source_;
+  WasmStorePtr store_;
+  WasmModulePtr module_;
+  WasmSharedModulePtr shared_module_;
+  WasmInstancePtr instance_;
+
+  WasmMemoryPtr memory_;
+  WasmTablePtr table_;
+
+  std::unordered_map<std::string, FuncDataPtr> host_functions_;
+  std::unordered_map<std::string, WasmFuncPtr> module_functions_;
+};
+
+// TODO(mathetake): move to proxy_wasm::common::*
+static uint32_t parseVarint(const byte_t *&pos, const byte_t *end) {
+  uint32_t n = 0;
+  uint32_t shift = 0;
+  byte_t b;
+  do {
+    if (pos + 1 > end) {
+      abort();
+    }
+    b = *pos++;
+    n += (b & 0x7f) << shift;
+    shift += 7;
+  } while ((b & 0x80) != 0);
+  return n;
+}
+
+bool Wasmtime::load(const std::string &code, bool allow_precompiled) {
+  store_ = wasm_store_new(engine());
+
+  // Wasm file header is 8 bytes (magic number + version).
+  static const uint8_t magic_number[4] = {0x00, 0x61, 0x73, 0x6d};
+  if (code.size() < 8 || ::memcmp(code.data(), magic_number, 4) != 0) {
+    return false;
+  }
+
+  wasm_byte_vec_new_uninitialized(source_.get(), code.size());
+  ::memcpy(source_.get()->data, code.data(), code.size());
+
+  WasmByteVec stripped;
+  wasm_byte_vec_new_empty(stripped.get());
+  module_ =
+      wasm_module_new(store_.get(), getStrippedSource(&stripped) ? stripped.get() : source_.get());
+
+  if (module_) {
+    shared_module_ = wasm_module_share(module_.get());
+    assert(shared_module_ != nullptr);
+  }
+
+  return module_ != nullptr;
+}
+
+std::unique_ptr<WasmVm> Wasmtime::clone() {
+  assert(shared_module_ != nullptr);
+  auto clone = std::make_unique<Wasmtime>();
+
+  clone->integration().reset(integration()->clone());
+  clone->store_ = wasm_store_new(engine());
+  clone->module_ = wasm_module_obtain(clone->store_.get(), shared_module_.get());
+  wasm_byte_vec_new_empty(clone->source_.get()); // to avoid freeing on nullptr in wasmtime
```

Comment:
This is a weird mix of raw pointers, smart pointers and deleters.

If we always need to initialize the vector (because we've added deleter to it, so it's always going to be destroyed), then perhaps it should be a smart pointer with default initializer or a class?

Reply:
agreed